pkg/ingester/limiting.go

```diff
@@ -0,0 +1,94 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package ingester
+
+import (
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/go-kit/log/level"
+	"github.com/shirou/gopsutil/v3/process"
+	"github.com/weaveworks/common/httpgrpc"
+)
+
+func (i *Ingester) updateResourceUtilization() {
+	if !i.cfg.UtilizationBasedLimitingEnabled {
+		return
+	}
+
+	lastUpdate := i.lastResourceUtilizationUpdate.Load()
+
+	proc, err := process.NewProcess(int32(os.Getpid()))
+	if err != nil {
+		level.Warn(i.logger).Log("msg", "couldn't update CPU/memory utilization - failed to get process info",
+			"err", err.Error())
+		return
+	}
+
+	memInfo, err := proc.MemoryInfo()
+	if err != nil {
+		level.Warn(i.logger).Log("msg", "couldn't update CPU/memory utilization - failed to get memory info",
+			"err", err.Error())
+		return
+	}
+	i.memoryUtilization.Store(memInfo.RSS)
+
+	now := time.Now().UTC()
+
+	cput, err := proc.Times()
+	if err != nil {
+		level.Warn(i.logger).Log("msg", "couldn't update CPU/memory utilization - failed to get process times",
+			"err", err.Error())
+		return
+	}
+
+	lastCPUTime := i.lastCPUTime.Load()
+	// TODO: Should Guest and GuestNice be subtracted, as potentially duplicated in User?
+	// https://github.com/mackerelio/go-osstat/blob/54b9216143f9aa087151258f103dbd5c381f7915/cpu/cpu_linux.go#L70-L74
+	cpuTime := cput.User + cput.System + cput.Idle + cput.Nice + cput.Iowait + cput.Irq +
+		cput.Softirq + cput.Steal + cput.Guest + cput.GuestNice
+	i.lastCPUTime.Store(cpuTime)
+
+	i.lastResourceUtilizationUpdate.Store(now)
+
+	if lastUpdate.IsZero() {
+		return
+	}
+
+	cpuUtil := (cpuTime - lastCPUTime) / now.Sub(lastUpdate).Seconds()
```

Comment:
For this PR, could we just fetch total CPU usage as a percentage rather than computing it ourselves?

Unfortunately, I don't think we can separate read vs write CPU utilization without changes to the go runtime, or thread pinning.

Reply:
Interesting, hadn't seen that. In any case, we likely won't be using gopsutil by the time the PR is finalized.

Reply:
Also, the [Percent](https://pkg.go.dev/github.com/shirou/gopsutil/v3@v3.23.4/cpu#Percent) function isn't suitable, since it calculates the system's CPU utilization and not that of a single process (we want to know how much CPU the ingester process is using).

Reply:
I see on the other hand that there's a [Process.Percent](https://pkg.go.dev/github.com/shirou/gopsutil/v3@v3.23.4/process#Process.Percent) method. I'll investigate its implementation.

Reply:
Having checked the `Process.Percent` implementation, it's essentially the same as mine AFAICT. Though, I wonder why the `cpu.Percent` function subtracts Guest and GuestNice from the total CPU time, while `process.Process.Percent` does not?? Guess it warrants some research.

Reply:
Yea, `cpu.Percent` would only be good to use if Mimir is the only container running in an ingester pod, which it seems to be, at least in [the Helm chart](https://github.com/grafana/mimir/blob/main/operations/helm/charts/mimir-distributed/templates/ingester/ingester-statefulset.yaml#L108).

Reply:
Yeah, it's generally alone in its pod. I've never experienced otherwise.

Reply:
(Note that since this comment was made, the PR has changed into measuring via cgroup v2 and cgroup v1 before it measures via the process' resource utilization.